File: apps/api/src/properties/properties.service.ts
import { HttpStatus, Injectable } from '@nestjs/common';
import { Property, PropertyType } from '@prisma/client';
import { PropertyDto, WS_EVENTS } from '@nodescope/shared';
import { PropertiesRepository } from './properties.repository';
import { ContainmentService } from './containment.service';
import { ConflictResolutionService } from '../conflict/conflict.service';
import { AuditService } from '../audit/audit.service';
import { NodeScopeException } from '../common/filters/global-exception.filter';
import { assertValidNesting } from './property-nesting';
import { CreatePropertyDto, PatchPropertyDto, PROPERTY_WRITABLE_FIELDS } from './properties.dto';
import type { OrgMemberContext } from '../organizations/org-context.types';
import { PermissionsService } from '../permissions/permissions.service';

@Injectable()
export class PropertiesService {
  constructor(
    private readonly repo: PropertiesRepository,
    private readonly conflict: ConflictResolutionService,
    private readonly audit: AuditService,
    private readonly containment: ContainmentService,
    private readonly permissions: PermissionsService,
  ) {}

  /** Org-scoped row lookup for cross-module callers (Spec 1 building-models). No permission-scope filter. */
  findInOrg(organizationId: string, id: string): Promise<Property | null> {
    return this.repo.findByIdAndOrgId(id, organizationId);
  }

  async listProperties(member: OrgMemberContext): Promise<PropertyDto[]> {
    const scope = await this.permissions.scopeFilter(member);
    return (await this.repo.findAllByOrgId(member.organizationId, scope)).map((p) => this.toDto(p));
  }

  async getProperty(member: OrgMemberContext, id: string): Promise<PropertyDto> {
    const scope = await this.permissions.scopeFilter(member);
    const p = await this.repo.findByIdAndOrgId(id, member.organizationId, scope);
    if (!p) throw new NodeScopeException('PROP_001', 'PROPERTY_NOT_FOUND', HttpStatus.NOT_FOUND);
    return this.toDto(p);
  }

  async createProperty(member: OrgMemberContext, dto: CreatePropertyDto): Promise<PropertyDto> {
    const organizationId = member.organizationId;
    // Top-level SITE: OWNER-only; sub-site: assertCanConfigure the parent
    if (dto.parentId == null) {
      await this.permissions.assertCanConfigure(member, '__root__');
    } else {
      await this.permissions.assertCanConfigure(member, dto.parentId);
    }

    const parentType = await this.resolveParentType(organizationId, dto.parentId ?? null);
    assertValidNesting(parentType, dto.type);
    if (await this.repo.existsSiblingName(organizationId, dto.parentId ?? null, dto.name)) {
      throw new NodeScopeException('PROP_003', 'PROPERTY_NAME_TAKEN', HttpStatus.CONFLICT);
    }
    const created = await this.repo.create({
      organizationId, parentId: dto.parentId ?? null, type: dto.type, name: dto.name, code: dto.code ?? null,
    });
    await this.conflict.emitScoped(organizationId, created.id, WS_EVENTS.PROPERTY_CREATED, { id: created.id });
    await this.audit.recordCreate(organizationId, 'Property', created);
    return this.toDto(created);
  }

  async updateProperty(member: OrgMemberContext, id: string, patch: PatchPropertyDto): Promise<PropertyDto> {
    const organizationId = member.organizationId;
    // Write path: org-wide lookup (no scope) so out-of-scope ADMIN → PERM_001 not 404
    const current = await this.repo.findByIdAndOrgId(id, organizationId);
    if (!current) throw new NodeScopeException('PROP_001', 'PROPERTY_NOT_FOUND', HttpStatus.NOT_FOUND);
    await this.permissions.assertCanConfigure(member, current.id);

    const changedField = (f: string) => patch.changes.find((c) => c.field === f);
    const parentChange = changedField('parentId');
    const nameChange = changedField('name');

    const nextParentId = parentChange ? ((parentChange.newValue as string | null) ?? null) : current.parentId;

    if (parentChange) {
      // If reparenting, also assert access to the destination
      if (nextParentId == null) {
        // Promoting to top-level is OWNER-only
        await this.permissions.assertCanConfigure(member, '__root__');
      } else {
        await this.permissions.assertCanConfigure(member, nextParentId);
      }

      // cycle: the new parent must not be the node itself or any of its descendants
      if (nextParentId) {
        if (nextParentId === id) throw new NodeScopeException('PROP_005', 'PROPERTY_CYCLE', HttpStatus.UNPROCESSABLE_ENTITY);
        const subtree = await this.repo.getSubtreeIds(organizationId, id);
        if (subtree.includes(nextParentId)) throw new NodeScopeException('PROP_005', 'PROPERTY_CYCLE', HttpStatus.UNPROCESSABLE_ENTITY);
      }
      const newParentType = await this.resolveParentType(organizationId, nextParentId);
      assertValidNesting(newParentType, current.type);
      await this.containment.assertReparentKeepsContainment(organizationId, id, nextParentId);
    }

    if (nameChange || parentChange) {
      const nextName = nameChange ? (nameChange.newValue as string) : current.name;
      if (await this.repo.existsSiblingName(organizationId, nextParentId, nextName, id)) {
        throw new NodeScopeException('PROP_003', 'PROPERTY_NAME_TAKEN', HttpStatus.CONFLICT);
      }
    }

    const payload = this.conflict.buildUpdatePayload(patch, PROPERTY_WRITABLE_FIELDS, current.version, CreatePropertyDto);
    const updated = await this.repo.updateWithVersion(id, organizationId, payload, patch.baseVersion);
    if (!updated) throw new NodeScopeException('SYNC_001', 'EDIT_CONFLICT', HttpStatus.CONFLICT);

    if (parentChange) {
      const oldParentId = current.parentId;
      await this.conflict.emitScopedMulti(
        organizationId,
        [id, oldParentId].filter((x): x is string => !!x),
        WS_EVENTS.PROPERTY_MOVED,
        { id },
      );
    } else {
      await this.conflict.emitScoped(organizationId, id, WS_EVENTS.PROPERTY_UPDATED, { id });
    }
    await this.audit.recordUpdate(organizationId, 'Property', id, patch.changes);
    return this.toDto(updated);
  }

  async deleteProperty(member: OrgMemberContext, id: string): Promise<void> {
    const organizationId = member.organizationId;
    // Write path: org-wide lookup (no scope) so out-of-scope ADMIN → PERM_001 not 404
    const p = await this.repo.findByIdAndOrgId(id, organizationId);
    if (!p) throw new NodeScopeException('PROP_001', 'PROPERTY_NOT_FOUND', HttpStatus.NOT_FOUND);
    await this.permissions.assertCanConfigure(member, p.id);
    const subtreeIds = await this.repo.getSubtreeIds(organizationId, id);
    const hasChildren = subtreeIds.length > 1;
    const devices = await this.repo.countDevicesUnder(organizationId, subtreeIds);
    const charters = await this.repo.countChartersUnder(organizationId, subtreeIds);
    if (hasChildren || devices > 0 || charters > 0) {
      throw new NodeScopeException('PROP_004', 'PROPERTY_NOT_EMPTY', HttpStatus.CONFLICT);
    }
    const assignments = await this.repo.countAssignmentsUnder(organizationId, subtreeIds);
    if (assignments > 0) {
      throw new NodeScopeException('PERM_005', 'PROPERTY_ASSIGNED', HttpStatus.CONFLICT);
    }
    // Emit BEFORE delete so the ancestor lookup in emitScoped can still resolve the row
    await this.conflict.emitScoped(organizationId, id, WS_EVENTS.PROPERTY_DELETED, { id });
    await this.repo.deleteByIdAndOrgId(id, organizationId);
    await this.audit.recordDelete(organizationId, 'Property', p);
  }

  // ---- helpers exposed for Phase B containment + F3 (spec §10.2) ----
  subtreePropertyIds(organizationId: string, id: string): Promise<string[]> {
    return this.repo.getSubtreeIds(organizationId, id);
  }
  isAtOrUnder(organizationId: string, descendantId: string, ancestorId: string): Promise<boolean> {
    return this.repo.isAtOrUnder(organizationId, descendantId, ancestorId);
  }

  private async resolveParentType(organizationId: string, parentId: string | null): Promise<PropertyType | null> {
    if (!parentId) return null;
    // Internal call: no scope — must see the full org tree for nesting validation
    const parent = await this.repo.findByIdAndOrgId(parentId, organizationId);
    if (!parent) throw new NodeScopeException('PROP_001', 'PROPERTY_NOT_FOUND', HttpStatus.NOT_FOUND);
    return parent.type;
  }

  private toDto(p: Property): PropertyDto {
    return {
      id: p.id, organizationId: p.organizationId, parentId: p.parentId, type: p.type,
      name: p.name, code: p.code, version: p.version,
      createdAt: p.createdAt.toISOString(), updatedAt: p.updatedAt.toISOString(),
    };
  }
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Application/Properties/PropertiesService.cs
using System.Text.Json;
using NodeScope.Contracts.Realtime;
using NodeScope.Modules.Inventory.Domain;
using NodeScope.Platform.Abstractions;

namespace NodeScope.Modules.Inventory.Application.Properties;

/// <summary>The property hierarchy's CRUD, ported from Node's <c>PropertiesService</c>.</summary>
public sealed class PropertiesService
{
    private readonly IPropertyRepository _properties;
    private readonly ContainmentService _containment;
    private readonly IPermissionScopeService _permissions;
    private readonly IRealtimeService _realtime;
    private readonly IAuditService _audit;

    public PropertiesService(
        IPropertyRepository properties,
        ContainmentService containment,
        IPermissionScopeService permissions,
        IRealtimeService realtime,
        IAuditService audit)
    {
        _properties = properties;
        _containment = containment;
        _permissions = permissions;
        _realtime = realtime;
        _audit = audit;
    }

    public async Task<IReadOnlyList<PropertyDto>> ListAsync(OrgMemberContext member, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var scope = await _permissions.ScopePropertyIdsAsync(member, cancellationToken);
        var rows = await _properties.ListAsync(member.OrganizationId, scope, cancellationToken);
        return [.. rows.Select(row => row.ToDto())];
    }

    public async Task<PropertyDto> GetAsync(OrgMemberContext member, string id, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var scope = await _permissions.ScopePropertyIdsAsync(member, cancellationToken);
        var row = await _properties.FindAsync(member.OrganizationId, id, scope, cancellationToken)
            ?? throw InventoryErrors.PropertyNotFound();
        return row.ToDto();
    }

    public async Task<PropertyDto> CreateAsync(
        OrgMemberContext member,
        CreatePropertyRequest request,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(request);
        var organizationId = member.OrganizationId;
        // Top-level SITE is OWNER-only ("__root__" is never in any admin's scope).
        await _permissions.AssertCanConfigureAsync(member, request.ParentId ?? "__root__", cancellationToken);

        var type = PropertyTypeLabels.TryParse(request.Type)!.Value;
        var parentType = await ResolveParentTypeAsync(organizationId, request.ParentId, cancellationToken);
        PropertyNesting.AssertValidNesting(parentType, type);
        if (await _properties.SiblingNameExistsAsync(organizationId, request.ParentId, request.Name!, null, cancellationToken))
        {
            throw InventoryErrors.PropertyNameTaken();
        }

        var created = await _properties.CreateAsync(
            new NewProperty(organizationId, request.ParentId, type, request.Name!, request.Code),
            cancellationToken);
        await _realtime.EmitScopedAsync(
            organizationId, created.Id, WsEvents.PropertyCreated, TimestampedId(created.Id), cancellationToken);
        await _audit.RecordCreateAsync(organizationId, "Property", created.Id, created.ToDto(), cancellationToken);
        return created.ToDto();
    }

    public async Task<PropertyDto> UpdateAsync(
        OrgMemberContext member,
        string id,
        ChangesetRequest patch,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(patch);
        var organizationId = member.OrganizationId;
        // Write path: org-wide lookup (no scope) so an out-of-scope ADMIN gets PERM_001, not 404.
        var current = await _properties.FindAsync(organizationId, id, null, cancellationToken)
            ?? throw InventoryErrors.PropertyNotFound();
        await _permissions.AssertCanConfigureAsync(member, current.Id, cancellationToken);

        var parentChange = patch.Changes?.FirstOrDefault(change => change.Field == "parentId");
        var nameChange = patch.Changes?.FirstOrDefault(change => change.Field == "name");
        var nextParentId = parentChange is null
            ? current.ParentId
            : ChangesetValues.AsString(parentChange.NewValue);

        if (parentChange is not null)
        {
            await _permissions.AssertCanConfigureAsync(member, nextParentId ?? "__root__", cancellationToken);
            if (nextParentId is not null
                && await _properties.IsAtOrUnderAsync(organizationId, nextParentId, id, cancellationToken))
            {
                throw InventoryErrors.PropertyCycle();
            }

            var newParentType = await ResolveParentTypeAsync(organizationId, nextParentId, cancellationToken);
            PropertyNesting.AssertValidNesting(newParentType, current.Type);
            await _containment.AssertReparentKeepsContainmentAsync(organizationId, id, nextParentId, cancellationToken);
        }

        if (nameChange is not null || parentChange is not null)
        {
            var nextName = nameChange is null ? current.Name : ChangesetValues.AsString(nameChange.NewValue) ?? "";
            if (await _properties.SiblingNameExistsAsync(organizationId, nextParentId, nextName, id, cancellationToken))
            {
                throw InventoryErrors.PropertyNameTaken();
            }
        }

        var payload = Changesets.BuildUpdatePayload(patch, PropertyFields.Writable, current.Version, PropertyFields.IsValid);
        var updated = await _properties.UpdateWithVersionAsync(
            organizationId, id, payload, patch.BaseVersion!.Value, cancellationToken)
            ?? throw Changesets.EditConflict();

        if (parentChange is not null)
        {
            string[] sites = current.ParentId is null ? [id] : [id, current.ParentId];
            await _realtime.EmitScopedMultiAsync(
                organizationId, sites, WsEvents.PropertyMoved, TimestampedId(id), cancellationToken);
        }
        else
        {
            await _realtime.EmitScopedAsync(
                organizationId, id, WsEvents.PropertyUpdated, TimestampedId(id), cancellationToken);
        }

        await _audit.RecordUpdateAsync(
            organizationId, "Property", id, ChangesetAudit.ToFieldChanges(patch), cancellationToken);
        return updated.ToDto();
    }

    public async Task DeleteAsync(OrgMemberContext member, string id, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var organizationId = member.OrganizationId;
        var property = await _properties.FindAsync(organizationId, id, null, cancellationToken)
            ?? throw InventoryErrors.PropertyNotFound();
        await _permissions.AssertCanConfigureAsync(member, property.Id, cancellationToken);

        var subtreeIds = await _properties.SubtreeIdsAsync(organizationId, id, cancellationToken);
        var hasChildren = subtreeIds.Count > 1;
        var devices = await _properties.CountDevicesUnderAsync(organizationId, subtreeIds, cancellationToken);
        var charters = await _properties.CountChartersUnderAsync(organizationId, subtreeIds, cancellationToken);
        if (hasChildren || devices > 0 || charters > 0)
        {
            throw InventoryErrors.PropertyNotEmpty();
        }

        if (await _properties.CountBuildingModelsUnderAsync(organizationId, subtreeIds, cancellationToken) > 0)
        {
            throw InventoryErrors.BuildingHasModel();
        }

        if (await _properties.CountAssignmentsUnderAsync(organizationId, subtreeIds, cancellationToken) > 0)
        {
            throw InventoryErrors.PropertyAssigned();
        }

        // Emit before the delete so scope resolution can still see the row.
        await _realtime.EmitScopedAsync(
            organizationId, id, WsEvents.PropertyDeleted, TimestampedId(id), cancellationToken);
        await _properties.DeleteAsync(organizationId, id, cancellationToken);
        await _audit.RecordDeleteAsync(organizationId, "Property", id, property.ToDto(), cancellationToken);
    }

    private async Task<PropertyType?> ResolveParentTypeAsync(
        string organizationId,
        string? parentId,
        CancellationToken cancellationToken)
    {
        if (parentId is null)
        {
            return null;
        }

        // Internal call, unscoped: nesting validation must see the full org tree.
        var parent = await _properties.FindAsync(organizationId, parentId, null, cancellationToken)
            ?? throw InventoryErrors.PropertyNotFound();
        return parent.Type;
    }

    private static object TimestampedId(string id) => new { id, timestamp = IsoTimestamp.Now() };
}
